packages/contracts/contracts/APM/Repo.sol

```diff
@@ -0,0 +1,224 @@
+    bytes32 public constant UPGRADE_ROLE = keccak256("UPGRADE_ROLE");
+
+    /// @notice Thrown version bump is invalid
+    error InvalidBump();
+
+    /// @notice Thrown if version is invalid
+    error InvalidVersion();
+
+    /// @notice Thrown if version does not exist
+    error InexistentVersion();
+
+    /// @notice Thrown if contract does not have IPluginFactory interface
+    error InvalidPluginInterface();
+
+    /// @notice Thrown if address is not a PluginFactory contract
+    error InvalidPluginContract();
+
+    /// @notice Thrown if address is not a contract
+    error InvalidContract();
+
+    struct Version {
+        uint16[3] semanticVersion;
+        address pluginFactoryAddress;
+        bytes contentURI;
+    }
+
+    uint256 internal versionsNextIndex = 0;
+    mapping(uint256 => Version) internal versions;
+    mapping(bytes32 => uint256) internal versionIdForSemantic;
+    mapping(address => uint256) internal latestVersionIdForContract;
+
+    event NewVersion(uint256 versionId, uint16[3] semanticVersion);
+
+    /// @dev Used for UUPS upgradability pattern
+    function initialize(address initialOwner) external initializer {
+        _registerStandard(type(IRepo).interfaceId);
+        __ACL_init(initialOwner);
+
+        versionsNextIndex = 1;
+
+        // set roles.
+        _grant(address(this), initialOwner, CREATE_VERSION_ROLE);
+    }
+
+    /// @dev Used to check the permissions within the upgradability pattern implementation of OZ
+    function _authorizeUpgrade(address)
+        internal
+        virtual
+        override
+        auth(address(this), UPGRADE_ROLE)
+    {}
+
+    /// @inheritdoc IRepo
+    function newVersion(
+        uint16[3] memory _newSemanticVersion,
+        address _pluginFactoryAddress,
+        bytes calldata _contentURI
+    ) external auth(address(this), CREATE_VERSION_ROLE) {
+        // check if _pluginFactoryAddress is IPluginFactory
+        if (Address.isContract(_pluginFactoryAddress)) {
+            try
+                IPluginFactory(_pluginFactoryAddress).supportsInterface(
+                    PluginFactoryIDs.PLUGIN_FACTORY_INTERFACE_ID
+                )
+            returns (bool result) {
+                if (!result) revert InvalidPluginInterface();
+            } catch {
+                revert InvalidPluginContract();
+            }
+        } else if (_pluginFactoryAddress != address(0)) revert InvalidContract();
+
+        address pluginFactoryAddress = _pluginFactoryAddress;
+        uint256 lastVersionIndex = versionsNextIndex - 1;
+
+        uint16[3] memory lastSematicVersion;
+
+        if (lastVersionIndex > 0) {
+            Version storage lastVersion = versions[lastVersionIndex];
+            lastSematicVersion = lastVersion.semanticVersion;
+
+            if (pluginFactoryAddress == address(0)) {
+                pluginFactoryAddress = lastVersion.pluginFactoryAddress;
+            }
```

Comment:
As said above, I would expect the address to either contain a new contract version (major) or to have the exact same address as before.